### PR TITLE
locationd: Add camera-IMU cross-checks

### PR DESCRIPTION
**Pull request by nuwandavek**
Problem: Erroneous gyro measurements cause a degradation in the localizer, which adversely affects driving. We should detect the bad measurements, do not update them to the filter, and finally add an alert.
image

To Do:
- [x]  Crosscheck gyro and cam odo rot to each other
- [x] Fix the decays to ensure sane alert duration
- [x] Run offline on a large seglist and compute TP/FP

### CODE REVIEW

**Comment by nuwandavek** [comment]
@adeebshihadeh I think we can have an alert (I checked on 50k segs, without any FP). But we could also maybe add without alert on `inputsOK`, track closely offline and then add it in the next release. Don't have strong opinions on it, but it should have the alert eventually for sure.

File: selfdrive/controls/controlsd.py [hunk 1]
```diff
@@ -378,7 +378,7 @@ def update_events(self, CS):
       self.events.add(EventName.vehicleModelInvalid)
     if not self.sm['lateralPlan'].mpcSolutionValid:
       self.events.add(EventName.plannerError)
-    if not (self.sm['liveParameters'].sensorValid or self.sm['liveLocationKalman'].sensorsOK) and not NOSENSOR:
+    if not (self.sm['liveParameters'].sensorValid or self.sm['liveLocationKalman'].sensorsOK or self.sm['liveLocationKalman'].inputsOK) and not NOSENSOR:
```
Comment (nuwandavek) [inline selfdrive/controls/controlsd.py:381]: `and` instead of `or`? It should trigger on all checks.

Comment (adeebshihadeh) [inline selfdrive/controls/controlsd.py:381]: `inputsOK` captures a lot more than sensors (e.g. `carState` going invalid). we need to be very careful about alert text; this seems like it deserves its own alert

**Review by adeebshihadeh — CHANGES_REQUESTED** [review]

File: selfdrive/locationd/locationd.cc [hunk 2]
```diff
@@ -711,9 +720,9 @@ int Localizer::locationd_thread() {
     // 100Hz publish for notcars, 20Hz for cars
     const char* trigger_msg = sm["carParams"].getCarParams().getNotCar() ? "accelerometer" : "cameraOdometry";
     if (sm.updated(trigger_msg)) {
-      bool inputsOK = sm.allAliveAndValid() && this->are_inputs_ok();
+      bool inputsOK = sm.allAliveAndValid();
       bool gpsOK = this->is_gps_ok();
-      bool sensorsOK = sm.allAliveAndValid({"accelerometer", "gyroscope"});
+      bool sensorsOK = sm.allAliveAndValid({"accelerometer", "gyroscope"}) && this->are_inputs_ok();
```
Comment (adeebshihadeh) [inline selfdrive/locationd/locationd.cc:725]: doesn't this include a bunch of non-sensor stuff?

Comment (nuwandavek) [reply selfdrive/locationd/locationd.cc:725]: It has `{"cameraOdometry", "liveCalibration", "accelerometer", "gyroscope"}`. 
All are necessary to be checked, IMO. Split this to sensor only for now?

**Comment by nuwandavek** [comment]
`excessiveResets` triggered `localizerMalfunction` alert, but did not do anything. In this PR, that alert has been repurposed to be the catch-all localizer alert, and `inputsOK` triggers it now.

File: selfdrive/controls/controlsd.py [hunk 3]
```diff
@@ -378,9 +378,11 @@ def update_events(self, CS):
       self.events.add(EventName.vehicleModelInvalid)
     if not self.sm['lateralPlan'].mpcSolutionValid:
       self.events.add(EventName.plannerError)
-    if not (self.sm['liveParameters'].sensorValid or self.sm['liveLocationKalman'].sensorsOK) and not NOSENSOR:
+    if not (self.sm['liveParameters'].sensorValid and self.sm['liveLocationKalman'].sensorsOK) and not NOSENSOR:
```
Comment (adeebshihadeh) [inline selfdrive/controls/controlsd.py:381]: not relevant to this PR, should be split out

**Pull request merged**